feat: support period vote tx

troggy opened the pull request:
- [x] Validators emit PeriodVote tx when it is time to submit period.
- [x] Validators collect period votes in a state 

Resolves #282

johannbarbie commented on src/period/submitPeriodVote.js:
```diff
@@ -0,0 +1,40 @@
+    bridgeState.account.privateKey
+  );
+
+  await sendDelayed(periodVoteTx);
```
why does this one need to be delayed? isn't tendermint deterministic? so if we got the period full, we are sure it will not go back?

troggy replied:
I have no answer to this yet, just used the existing approach: https://github.com/leapdao/leap-node/blob/master/src/eventsRelay.js#L65

johannbarbie commented on src/tx/applyTx/checkPeriodVote.js:
```diff
@@ -0,0 +1,68 @@
+
+  // validators are not in consensus on period roots atm
+  // disabling this until resolved
+  // https://github.com/leapdao/leap-node/issues/282#issuecomment-519160276
```
issue solved, should this be reenabled?

troggy replied:
it is still not working right. I'm fixing it as part of https://github.com/leapdao/leap-node/issues/309 here: https://github.com/leapdao/leap-node/tree/feat/submit-with-cas

TLDR: bad idea to submit period for tx handler. I'm submitting period for block middleware now, PR to follow once done

johannbarbie commented on src/period/index.js:
```diff
@@ -19,8 +24,14 @@ module.exports = (bridgeState, nodeConfig = {}) => async (rsp, chainInfo, chainH
-  await new Promise((resolve) => {
+  await new Promise(resolve => {
     setTimeout(async () => {
+      await submitPeriodVote(
```
why did you decide to call `submitPeriodVote()` here?

this is called when tendermint is trying to propose a new block (right?), and `submitPeriod()` is called as a fallback, in case the selected proposer is offline.

troggy replied:
this handler is called every period (`checkPeriod` middleware). Right, I'm submitting the vote here as a fallback as well (`submitPeriodVote` has a protection against double votes).

`updatePeriod` is being called every block and that's where I submit a vote first: https://github.com/leapdao/leap-node/pull/305/files/15924f654aabe89a4492c03fa3c51f7f67b02545#diff-786c9ee7cb87a989994653f4a23150a6R25

johannbarbie commented on src/txHelpers/submitPeriod.test.js:
```diff
@@ -180,6 +214,121 @@ describe('submitPeriod', () => {
+    expect(submitPeriodWithCas).not.toBeCalled();
   });
+
+  // describe('period vote', () => {
```
no code comments

johannbarbie commented on src/tx/applyTx/checkPeriodVote.js:
```diff
@@ -0,0 +1,68 @@
+    );
+  }
+
+  const periodRoot = bufferToHex(tx.inputs[0].prevout.hash);
```
also make sure that `tx.inputs[0].prevout.pos == 0`, otherwise it would screw up the index later.

johannbarbie commented on src/tx/applyTx/checkPeriodVote.js:
```diff
@@ -0,0 +1,68 @@
+  // validators are not in consensus on period roots atm
+  // disabling this until resolved
+  // https://github.com/leapdao/leap-node/issues/282#issuecomment-519160276
+  // const { result } = checkEnoughVotes(periodRoot, state);
```
remove code comments

troggy commented:
addressed everything.

+ refactored to keep periodVotes in a tendermint `state`, not in a `bridgeState`. This is the right way to do ([see](https://leapdao.slack.com/archives/CAAGW85BR/p1565270194076200?thread_ts=1565264683.074200&cid=CAAGW85BR))

johannbarbie commented on src/tx/applyTx/checkPeriodVote.test.js:
```diff
@@ -0,0 +1,161 @@
+  ...attrs,
+});
+
+describe('checkPeriodVote', () => {
```
❤️

johannbarbie commented on src/period/submitPeriodVote.test.js:
```diff
@@ -0,0 +1,96 @@
+  prevHash: '0x5678',
+};
+
+describe('submit period vote', () => {
```
❤️
